glean-core/src/upload/mod.rs

```diff
@@ -1136,4 +1182,44 @@ mod test {
             }
         }
     }
+
+    #[test]
+    fn quota_is_enforced_when_enqueueing_cached_pings() {
+        let (mut glean, tmpdir) = new_glean(None);
+
+        // Register a ping for testing
+        let ping_type = PingType::new("test", true, /* send_if_empty */ true, vec![]);
+        glean.register_ping_type(&ping_type);
+
+        // Submit the ping multiple times
+        let n = 10;
+        for _ in 0..n {
+            glean.submit_ping(&ping_type, None).unwrap();
+        }
+
+        // Create a new upload manager pointing to the same data_path as the glean instance.
+        let upload_manager = PingUploadManager::new(tmpdir.path(), "Rust", true);
+
+        // Enqueue cached pings and set the quota to just a little over the size on an empty ping file.
+        // This way we can check that one ping is kept and all others are deleted.
+        //
+        // From manual testing I figured out an empty ping file is 324bytes,
+        // I am setting this a little over just so that minor changes to the ping structure
+        // don't mmediatelly break this.
+        upload_manager.enqueue_cached_pings(&glean, 500);
+
+        // Get a task once
+        // One ping should have been enqueued.
+        match upload_manager.get_upload_task(&glean, false) {
+            PingUploadTask::Upload(_) => {}
```

Comment:
Hm, we _could_ check that we're seeing the last written ping (seq_id 9), but that would require us to parse the body into JSON, then look into that.
(or do some ugly string check).

Did you by chance check that it is the latest ping, just to verify our own assumptions in which order we delete?

Reply:
Hm, really good that you asked for this. Actually found a bug in the implementation where we are deleting the newest pings, not the oldest. Somehow we will need to "reverse drain" the cached_pings. Something for tomorrow's Bea.

Reply:
Hah, good I asked :D 
In that case let's make sure to encode this into the test after you fix it.
Guess we need 2 runs through the list then? 
One to trim it down to the maximum size, then one to insert in the right order?

Reply:
Also glad you asked! And you are correct, we need two run throughs and we need to reverse the vector for the first run thorugh and then reverse it back for enqueueing... I guess this will need another pass before I am confident to merge. You can look at the changes in this commit https://github.com/mozilla/glean/pull/1110/commits/586d84e628e4c1280d9c66e23a8f95eb289e5f37.